adapter/internal/ingress/client_manager.go

```diff
@@ -64,23 +65,29 @@ func NewClientManager(connector ConnectionBuilder, connCount int, ttl time.Durat
 func (c *ClientManager) Next() v2.EgressClient {
 	for {
 		idx := atomic.AddUint64(&c.nextIdx, 1)
-		conn := (*connection)(atomic.LoadPointer(&c.connections[idx%uint64(len(c.connections))]))
-		if conn.client != nil {
+		actualIdx := int(idx % uint64(len(c.connections)))
+
+		c.mu.RLock()
+		conn := c.connections[actualIdx]
+		if conn != nil && conn.client != nil {
 			return conn.client
 		}
+		c.mu.RUnlock()
```

Comment:
You are holding the read lock after you return and do not unlock. Why not just do:

```go
c.mu.RLock()
conn := c.connections[actualIdx]
c.mu.RUnlock()
```